vscode/src/azure/auth.ts

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import * as vscode from "vscode";
+
+import { log } from "qsharp-lang";
+import { EventType, sendTelemetryEvent, UserFlowStatus } from "../telemetry";
+import { getRandomGuid } from "../utils";
+import { azureRequest, AzureUris, ResponseTypes } from "./networkRequests";
+
+export const scopes = {
+  armMgmt: "https://management.azure.com/user_impersonation",
+  quantum: "https://quantum.microsoft.com/user_impersonation",
+};
+
+export async function getAuthSession(
+  scopes: string[],
+  associationId: string,
+): Promise<vscode.AuthenticationSession> {
+  const start = performance.now();
+  sendTelemetryEvent(EventType.AuthSessionStart, { associationId }, {});
+  log.debug("About to getSession for scopes", scopes.join(","));
+  try {
+    let session = await vscode.authentication.getSession("microsoft", scopes, {
+      silent: true,
+    });
+    if (!session) {
+      log.debug("No session with silent request. Trying with createIfNone");
+      session = await vscode.authentication.getSession("microsoft", scopes, {
+        createIfNone: true,
+      });
+    }
+    log.debug("Got session: ", JSON.stringify(session, null, 2));
+    sendTelemetryEvent(
+      EventType.AuthSessionEnd,
+      {
+        associationId,
+        flowStatus: UserFlowStatus.Succeeded,
+      },
+      { timeToCompleteMs: performance.now() - start },
+    );
+    return session;
+  } catch (e) {
+    sendTelemetryEvent(
+      EventType.AuthSessionEnd,
+      {
+        associationId,
+        reason: "exception in getAuthSession",
+        flowStatus: UserFlowStatus.Failed,
+      },
+      { timeToCompleteMs: performance.now() - start },
+    );
+    log.error("Exception occurred in getAuthSession: ", e);
+    throw e;
+  }
+}
+
+export async function getTenantIdAndToken(
+  associationId: string,
+  start: number,
```

Comment:
I really don't like flowing these args around just for telemetry properties. Will figure out a better approach here.